rct_optimizations/test/include/rct_optimizations_tests/pose_generator.h

```diff
@@ -25,70 +26,258 @@ struct PoseGenerator
    * @param target_origin The position of the target
    * @return A vector of camera positions & orientations
    */
-  virtual std::vector<Eigen::Isometry3d> generate(const Eigen::Vector3d &target_origin) const = 0;
+  virtual std::vector<Eigen::Isometry3d> generate(const Eigen::Isometry3d &target_origin) = 0;
+
+  virtual double getZRot() = 0;
 };
 
 /**
  * @brief Generates camera poses in a hemisphere pattern
  */
 struct HemispherePoseGenerator : PoseGenerator
 {
-  inline HemispherePoseGenerator(double r_ = 2.0,
-                                 unsigned theta_cnt_ = 10,
-                                 unsigned phi_cnt_ = 10)
+protected:
+  inline HemispherePoseGenerator(double r_,
+                                 unsigned theta_cnt_,
+                                 unsigned phi_cnt_,
+                                 double z_rot_min_,
+                                 double z_rot_max_,
+                                 const Eigen::Isometry3d& target_offset_,
+                                 unsigned long seed_ = std::mt19937::default_seed)
     : r(r_)
     , theta_cnt(theta_cnt_)
     , phi_cnt(phi_cnt_)
+    , target_offset(target_offset_)
+    , mt_rand(seed_)
+    , z_sampler(z_rot_min_, z_rot_max_)
+  {
+
+  }
```

Comment:
It seems a bit backwards to put the implementation specific details of the random Z rotation pose generator in the base class. The generic pose generator classes can be used legitimately without random engines and bounds.

To me it would make more sense if we left `HemispherePoseGenerator` as-is, and added the random engine and bounds to the `RandomZRotHemispherePoseGenerator`. The derived class (random Z rotation) could get the nominal hemisphere poses from the base class, and then apply the random rotations to each pose afterwards

Reply:
I structured it this way in an effort to minimize the amount of duplicate code I was adding. The public constructors of `HemispherePoseGenerator` set ranges for the Z rotation sampler so that it's locked to either zero or a constant value. The constructors for `RandomZRotHemispherePoseGenerator` set the rotation sampler bounds to `0.0` and `2*M_PI` through the protected constructor of `HemispherePoseGenerator`, which enables the random sampling. This way the random sampling pose generators can use the same `generate` function as the base class, and the only thing that is added is the new set of constructors..

Reply:
You could probably just use a single class that randomizes the Z-axis orientation for any input "nominal" pose generator

```c++
struct RandomZAxisRotation : PoseGenerator
{
  RandomZAxisRotation(std::shared_ptr<PoseGenerator> generator, double min, double max, double seed)
    : gen(generator)
    , mt_rand(seed)
    , z_sampler(min, max)
  {
  }

  std::vector<Eigen::Isometry3d> generate()
  {
    // Create the nominal poses
    auto poses = gen.generate();

    // Rotate each pose randomly about it's Z-axis
    for(auto& pose : out)
    {
      pose.rotate(Eigen::AngleAxisd(z_sampler(mt_rand), Eigen::Vector3d::UnitZ()));
    }
    return out;
  }

  ...

};
```

Reply:
That's a much more flexible way to do it.

Reply:
Revised to implement randomization the way you describe